share/man/man4/unix.4
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)unix.4	8.1 (Berkeley) 6/9/93
.\" $FreeBSD: head/share/man/man4/unix.4 301597 2016-06-08 10:26:17Z trasz $
.\"
.Dd March 19, 2013
.Dt UNIX 4
.Os
.Sh NAME
.Nm unix
.Nd UNIX-domain protocol family
.Sh SYNOPSIS
.In sys/types.h
.In sys/un.h
.Sh DESCRIPTION
The
.Ux Ns -domain
protocol family is a collection of protocols
that provides local (on-machine) interprocess
communication through the normal
.Xr socket 2
mechanisms.
The
.Ux Ns -domain
family supports the
.Dv SOCK_STREAM ,
.Dv SOCK_SEQPACKET ,
and
.Dv SOCK_DGRAM
socket types and uses
file system pathnames for addressing.
.Sh ADDRESSING
.Ux Ns -domain
addresses are variable-length file system pathnames of
at most 104 characters.
The include file
.In sys/un.h
defines this address:
.Bd -literal -offset indent
struct sockaddr_un {
	u_char	sun_len;
	u_char	sun_family;
	char	sun_path[104];
};
.Ed
.Pp
Binding a name to a
.Ux Ns -domain
socket with
.Xr bind 2
causes a socket file to be created in the file system.
This file is
.Em not
removed when the socket is closed \(em
.Xr unlink 2
must be used to remove the file.
.Pp
The length of
.Ux Ns -domain
address, required by
.Xr bind 2
and
.Xr connect 2 ,
can be calculated by the macro
.Fn SUN_LEN
defined in
.In sys/un.h .
The
.Va sun_path
field must be terminated by a
.Dv NUL
character to be used with
.Fn SUN_LEN ,
but the terminating
.Dv NUL
is
.Em not
part of the address.
.Pp
The
.Ux Ns -domain
protocol family does not support broadcast addressing or any form
of
.Dq wildcard
matching on incoming messages.
All addresses are absolute- or relative-pathnames
of other
.Ux Ns -domain
sockets.
Normal file system access-control mechanisms are also
applied when referencing pathnames; e.g., the destination
of a
.Xr connect 2
or
.Xr sendto 2
must be writable.
.Sh PASSING FILE DESCRIPTORS
The
.Ux Ns -domain
sockets support the communication of
.Ux
file descriptors through the use of the
.Va msg_control
field in the
.Fa msg
argument to
.Xr sendmsg 2
and
.Xr recvmsg 2 .
.Pp
Any valid descriptor may be sent in a message.
The file descriptor(s) to be passed are described using a
.Vt "struct cmsghdr"
that is defined in the include file
.In sys/socket.h .
The type of the message is
.Dv SCM_RIGHTS ,
and the data portion of the messages is an array of integers
representing the file descriptors to be passed.
The number of descriptors being passed is defined
by the length field of the message;
the length field is the sum of the size of the header
plus the size of the array of file descriptors.
.Pp
The received descriptor is a
.Em duplicate
of the sender's descriptor, as if it were created via
.Li dup(fd)
or
.Li fcntl(fd, F_DUPFD_CLOEXEC, 0)
depending on whether
.Dv MSG_CMSG_CLOEXEC
is passed in the
.Xr recvmsg 2
call.
Descriptors that are awaiting delivery, or that are
purposely not received, are automatically closed by the system
when the destination socket is closed.
.Sh SOCKET OPTIONS
.Tn UNIX
domain sockets support a number of socket options which can be set with
.Xr setsockopt 2
and tested with
.Xr getsockopt 2 :
.Bl -tag -width ".Dv LOCAL_CONNWAIT"
.It Dv LOCAL_CREDS
This option may be enabled on
.Dv SOCK_DGRAM ,
.Dv SOCK_SEQPACKET ,
or a
.Dv SOCK_STREAM
socket.
This option provides a mechanism for the receiver to
receive the credentials of the process as a
.Xr recvmsg 2
control message.
The
.Va msg_control
field in the
.Vt msghdr
structure points to a buffer that contains a
.Vt cmsghdr
structure followed by a variable length
.Vt sockcred
structure, defined in
.In sys/socket.h
as follows:
.Bd -literal
struct sockcred {
  uid_t	sc_uid;		/* real user id */
  uid_t	sc_euid;	/* effective user id */
  gid_t	sc_gid;		/* real group id */
  gid_t	sc_egid;	/* effective group id */
  int	sc_ngroups;	/* number of supplemental groups */
  gid_t	sc_groups[1];	/* variable length */
};
.Ed
.Pp
The
.Fn SOCKCREDSIZE
macro computes the size of the
.Vt sockcred
structure for a specified number
of groups.
The
.Vt cmsghdr
fields have the following values:
.Bd -literal
cmsg_len = CMSG_LEN(SOCKCREDSIZE(ngroups))
cmsg_level = SOL_SOCKET
cmsg_type = SCM_CREDS
.Ed
.Pp
On
.Dv SOCK_STREAM
and
.Dv SOCK_SEQPACKET
sockets credentials are passed only on the first read from a socket,
then system clears the option on socket.
.It Dv LOCAL_CONNWAIT
Used with
.Dv SOCK_STREAM
sockets, this option causes the
.Xr connect 2
function to block until
.Xr accept 2
has been called on the listening socket.
.It Dv LOCAL_PEERCRED
Requested via
.Xr getsockopt 2
on a
.Dv SOCK_STREAM
socket returns credentials of the remote side.
These will arrive in the form of a filled in
.Vt xucred
structure, defined in
.In sys/ucred.h
as follows:
.Bd -literal
struct xucred {
  u_int	cr_version;		/* structure layout version */
  uid_t	cr_uid;			/* effective user id */
  short	cr_ngroups;		/* number of groups */
  gid_t	cr_groups[XU_NGROUPS];	/* groups */
};
.Ed
The
.Vt cr_version
fields should be checked against
.Dv XUCRED_VERSION
define.
.Pp
The credentials presented to the server (the
.Xr listen 2
caller) are those of the client when it called
.Xr connect 2 ;
the credentials presented to the client (the
.Xr connect 2
caller) are those of the server when it called
.Xr listen 2 .
This mechanism is reliable; there is no way for either party to influence
the credentials presented to its peer except by calling the appropriate
system call (e.g.,
.Xr connect 2
or
.Xr listen 2 )
under different effective credentials.
.Pp
To reliably obtain peer credentials on a
.Dv SOCK_DGRAM
socket refer to the
.Dv LOCAL_CREDS
socket option.
.El
.Sh SEE ALSO
.Xr connect 2 ,
.Xr dup 2 ,
.Xr fcntl 2 ,
.Xr getsockopt 2 ,
.Xr listen 2 ,
.Xr recvmsg 2 ,
.Xr sendto 2 ,
.Xr setsockopt 2 ,
.Xr socket 2 ,
.Xr intro 4
.Rs
.%T "An Introductory 4.3 BSD Interprocess Communication Tutorial"
.%B PS1
.%N 7
.Re
.Rs
.%T "An Advanced 4.3 BSD Interprocess Communication Tutorial"
.%B PS1
.%N 8
.Re


usr.bin/netstat/unix.c
/*-
 * Copyright (c) 1983, 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if 0
#ifndef lint
static char sccsid[] = "@(#)unix.c	8.1 (Berkeley) 6/6/93";
#endif /* not lint */
#endif

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: head/usr.bin/netstat/unix.c 298182 2016-04-18 05:46:18Z araujo $");

/*
 * Display protocol blocks in the unix domain.
 */
#include <sys/param.h>
#include <sys/queue.h>
#include <sys/protosw.h>
#include <sys/socket.h>
#include <sys/socketvar.h>
#include <sys/mbuf.h>
#include <sys/sysctl.h>
#include <sys/un.h>
#include <sys/unpcb.h>

#include <netinet/in.h>

#include <errno.h>
#include <err.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <strings.h>
#include <kvm.h>
#include <libxo/xo.h>
#include "netstat.h"

static	void unixdomainpr(struct xunpcb *, struct xsocket *);

static	const char *const socktype[] =
    { "#0", "stream", "dgram", "raw", "rdm", "seqpacket" };

static int
pcblist_sysctl(int type, char **bufp)
{
	char 	*buf;
	size_t	len;
	char mibvar[sizeof "net.local.seqpacket.pcblist"];

	sprintf(mibvar, "net.local.%s.pcblist", socktype[type]);

	len = 0;
	if (sysctlbyname(mibvar, 0, &len, 0, 0) < 0) {
		if (errno != ENOENT)
			xo_warn("sysctl: %s", mibvar);
		return (-1);
	}
	if ((buf = malloc(len)) == NULL) {
		xo_warnx("malloc %lu bytes", (u_long)len);
		return (-2);
	}
	if (sysctlbyname(mibvar, buf, &len, 0, 0) < 0) {
		xo_warn("sysctl: %s", mibvar);
		free(buf);
		return (-2);
	}
	*bufp = buf;
	return (0);
}

static int
pcblist_kvm(u_long count_off, u_long gencnt_off, u_long head_off, char **bufp)
{
	struct unp_head head;
	struct unpcb *unp, unp_conn;
	u_char sun_len;
	struct socket so;
	struct xunpgen xug;
	struct xunpcb xu;
	unp_gen_t unp_gencnt;
	u_int	unp_count;
	char 	*buf, *p;
	size_t	len;

	if (count_off == 0 || gencnt_off == 0)
		return (-2);
	if (head_off == 0)
		return (-1);
	kread(count_off, &unp_count, sizeof(unp_count));
	len = 2 * sizeof(xug) + (unp_count + unp_count / 8) * sizeof(xu);
	if ((buf = malloc(len)) == NULL) {
		xo_warnx("malloc %lu bytes", (u_long)len);
		return (-2);
	}
	p = buf;

#define	COPYOUT(obj, size) do {						\
	if (len < (size)) {						\
		xo_warnx("buffer size exceeded");			\
		goto fail;						\
	}								\
	bcopy((obj), p, (size));					\
	len -= (size);							\
	p += (size);							\
} while (0)

#define	KREAD(off, buf, len) do {					\
	if (kread((uintptr_t)(off), (buf), (len)) != 0)			\
		goto fail;						\
} while (0)

	/* Write out header. */
	kread(gencnt_off, &unp_gencnt, sizeof(unp_gencnt));
	xug.xug_len = sizeof xug;
	xug.xug_count = unp_count;
	xug.xug_gen = unp_gencnt;
	xug.xug_sogen = 0;
	COPYOUT(&xug, sizeof xug);

	/* Walk the PCB list. */
	xu.xu_len = sizeof xu;
	KREAD(head_off, &head, sizeof(head));
	LIST_FOREACH(unp, &head, unp_link) {
		xu.xu_unpp = unp;
		KREAD(unp, &xu.xu_unp, sizeof (*unp));
		unp = &xu.xu_unp;

		if (unp->unp_gencnt > unp_gencnt)
			continue;
		if (unp->unp_addr != NULL) {
			KREAD(unp->unp_addr, &sun_len, sizeof(sun_len));
			KREAD(unp->unp_addr, &xu.xu_addr, sun_len);
		}
		if (unp->unp_conn != NULL) {
			KREAD(unp->unp_conn, &unp_conn, sizeof(unp_conn));
			if (unp_conn.unp_addr != NULL) {
				KREAD(unp_conn.unp_addr, &sun_len,
				    sizeof(sun_len));
				KREAD(unp_conn.unp_addr, &xu.xu_caddr, sun_len);
			}
		}
		KREAD(unp->unp_socket, &so, sizeof(so));
		if (sotoxsocket(&so, &xu.xu_socket) != 0)
			goto fail;
		COPYOUT(&xu, sizeof(xu));
	}

	/* Reread the counts and write the footer. */
	kread(count_off, &unp_count, sizeof(unp_count));
	kread(gencnt_off, &unp_gencnt, sizeof(unp_gencnt));
	xug.xug_count = unp_count;
	xug.xug_gen = unp_gencnt;
	COPYOUT(&xug, sizeof xug);

	*bufp = buf;
	return (0);

fail:
	free(buf);
	return (-1);
#undef COPYOUT
#undef KREAD
}

void
unixpr(u_long count_off, u_long gencnt_off, u_long dhead_off, u_long shead_off,
    u_long sphead_off, bool *first)
{
	char 	*buf;
	int	ret, type;
	struct	xsocket *so;
	struct	xunpgen *xug, *oxug;
	struct	xunpcb *xunp;
	u_long	head_off;

	buf = NULL;
	for (type = SOCK_STREAM; type <= SOCK_SEQPACKET; type++) {
		if (live)
			ret = pcblist_sysctl(type, &buf);
		else {
			head_off = 0;
			switch (type) {
			case SOCK_STREAM:
				head_off = shead_off;
				break;

			case SOCK_DGRAM:
				head_off = dhead_off;
				break;

			case SOCK_SEQPACKET:
				head_off = sphead_off;
				break;
			}
			ret = pcblist_kvm(count_off, gencnt_off, head_off,
			    &buf);
		}
		if (ret == -1)
			continue;
		if (ret < 0)
			return;

		oxug = xug = (struct xunpgen *)buf;
		for (xug = (struct xunpgen *)((char *)xug + xug->xug_len);
		    xug->xug_len > sizeof(struct xunpgen);
		    xug = (struct xunpgen *)((char *)xug + xug->xug_len)) {
			xunp = (struct xunpcb *)xug;
			so = &xunp->xu_socket;

			/* Ignore PCBs which were freed during copyout. */
			if (xunp->xu_unp.unp_gencnt > oxug->xug_gen)
				continue;
			if (*first) {
				xo_open_list("socket");
				*first = false;
			}
			xo_open_instance("socket");
			unixdomainpr(xunp, so);
			xo_close_instance("socket");
		}
		if (xug != oxug && xug->xug_gen != oxug->xug_gen) {
			if (oxug->xug_count > xug->xug_count) {
				xo_emit("Some {:type/%s} sockets may have "
				    "been {:action/deleted}.\n",
				    socktype[type]);
			} else if (oxug->xug_count < xug->xug_count) {
				xo_emit("Some {:type/%s} sockets may have "
				    "been {:action/created}.\n",
				    socktype[type]);
			} else {
				xo_emit("Some {:type/%s} sockets may have "
				    "been {:action/created or deleted}",
				    socktype[type]);
			}
		}
		free(buf);
	}
}

static void
unixdomainpr(struct xunpcb *xunp, struct xsocket *so)
{
	struct unpcb *unp;
	struct sockaddr_un *sa;
	static int first = 1;
	char buf1[33];
	static const char *titles[2] = {
	    "{T:/%-8.8s} {T:/%-6.6s} {T:/%-6.6s} {T:/%-6.6s} {T:/%8.8s} "
	    "{T:/%8.8s} {T:/%8.8s} {T:/%8.8s} {T:Addr}\n",
	    "{T:/%-16.16s} {T:/%-6.6s} {T:/%-6.6s} {T:/%-6.6s} {T:/%16.16s} "
	    "{T:/%16.16s} {T:/%16.16s} {T:/%16.16s} {T:Addr}\n"
	};
	static const char *format[2] = {
	    "{q:address/%8lx} {t:type/%-6.6s} "
	    "{:receive-bytes-waiting/%6u} "
	    "{:send-bytes-waiting/%6u} "
	    "{q:vnode/%8lx} {q:connection/%8lx} "
	    "{q:first-reference/%8lx} {q:next-reference/%8lx}",
	    "{q:address/%16lx} {t:type/%-6.6s} "
	    "{:receive-bytes-waiting/%6u} "
	    "{:send-bytes-waiting/%6u} "
	    "{q:vnode/%16lx} {q:connection/%16lx} "
	    "{q:first-reference/%16lx} {q:next-reference/%16lx}"
	};
	int fmt = (sizeof(void *) == 8) ? 1 : 0;

	unp = &xunp->xu_unp;
	if (unp->unp_addr)
		sa = &xunp->xu_addr;
	else
		sa = (struct sockaddr_un *)0;

	if (first && !Lflag) {
		xo_emit("{T:Active UNIX domain sockets}\n");
		xo_emit(titles[fmt],
		    "Address", "Type", "Recv-Q", "Send-Q",
		    "Inode", "Conn", "Refs", "Nextref");
		first = 0;
	}

	if (Lflag && so->so_qlimit == 0)
		return;

	if (Lflag) {
		snprintf(buf1, sizeof buf1, "%u/%u/%u", so->so_qlen,
		    so->so_incqlen, so->so_qlimit);
		xo_emit("unix  {d:socket/%-32.32s}{e:queue-length/%u}"
		    "{e:incomplete-queue-length/%u}{e:queue-limit/%u}",
		    buf1, so->so_qlen, so->so_incqlen, so->so_qlimit);
	} else {
		xo_emit(format[fmt],
		    (long)so->so_pcb, socktype[so->so_type], so->so_rcv.sb_cc,
		    so->so_snd.sb_cc, (long)unp->unp_vnode,
		    (long)unp->unp_conn,
		    (long)LIST_FIRST(&unp->unp_refs),
		    (long)LIST_NEXT(unp, unp_reflink));
	}
	if (sa)
		xo_emit(" {:path/%.*s}",
		    (int)(sa->sun_len - offsetof(struct sockaddr_un, sun_path)),
		    sa->sun_path);
	xo_emit("\n");
}
